5주차 레스토랑 조회 구현하기 PR 요청합니다.

Ubermensch0608 opened the pull request:
안녕하세요 트레이너님!

아직 완성하지 못 했지만 레스토랑 조회 구현하기 PR올립니다. 
단계적으로 앱을 만들어 보면서 커밋, TDD, redux 등 익숙해지고 싶어서 우선 커밋과 TDD에 집중해보았습니다.
기능적으로는 문제 없이 동작하고 App에서 모든 로직을 다루고 있습니다.  

이후에 redux를 이용한 상태관리 및 redux-thunk 적용할 계획입니다. 
감사합니다!

Ubermensch0608 committed:
RegionsContainer 테스트 코드 작성

지역 목록이 있을 경우와 없는 경우를 나누어 각각의 동작 결과를 예상하였습니다.

Ubermensch0608 committed:
지역 리스트 제공 유무에 따른 렌더링 테스트 구현

지역 리스트 배열이 비어있을 때와 한 개 이상 있을 때를 나누어 렌더링이 어떻게 다르게 나타나는 지 테스트하였습니다.

Ubermensch0608 committed:
카테고리 컴포넌트 렌더링 테스트

카테고리 리스트가 주어졌을 때와 그렇지 않을 때를 나누어 렌더링 테스트 하였습니다.

Ubermensch0608 committed:
레스토랑 목록 조회 테스트 구현

체크된 지역과 카테고리에 따라 조회된 레스토랑 목록을 화면에 구현하는지 확인하는 테스르 구현했습니다.

Ubermensch0608 committed:
Refactor: 데이터 리스트 api 호출 완료

좀 더 실제적인 구현을 위해 mock data에서 api호출로 리팩터링하였습니다.
카테고리와 지역 기반으로 레스토랑 데이터를 받을 수 있습니다.

daadaadaah commented on src/Category.test.jsx:
```diff
@@ -0,0 +1,18 @@
+import Category from './Category';
+
+describe('Category', () => {
+  describe('checkCategory', () => {
+    it('액션이 실행된다.', () => {
```
it문 하나로 표현하면 테스트의 의미가 달라질까요?

Ubermensch0608 replied:
말씀해주신 점 고민해보고 describe 중첩이 필요 없다고 판단하여 수정했습니다!

daadaadaah commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,28 @@
+
+import Regions from './Regions';
+
+describe('Regions', () => {
+  context('지역 리스트가 있을 때', () => {
+    it('리스트는 렌더한다.', () => {
```
describe-context-it 문에 있는 문장을 하나의 문장으로 했을 때, 자연스럽게 작성하는 것이 좋습니다.
지금 작성해주신대로 한문장을 만들면, `Region은 지역 리스트가 있을 때 리스트는 렌더한다.`라는 문장이 만들어지는데, 어떤가요? 어색하다면 어떻게 표현하는게 좋을까요? (다른 테스트도 한번 살펴봐주세요!)

Ubermensch0608 replied:
다시 보니 완전한 문장으로 느껴지지 않아 전반적으로 설명문 수정해봣습니다 :)

daadaadaah commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,28 @@
+describe('Regions', () => {
+  context('지역 리스트가 있을 때', () => {
+    it('리스트는 렌더한다.', () => {
+      const regions = [
+        { id: 1, regions: '서울' },
+        { id: 2, regions: '인천' },
+      ];
```
`regions`이라는 데이터는 다른 테스트에서도 사용되고 있어요. 이런 데이터들을 `fixture`라고 합니다. 이런 데이터들을 `fixtures`라는 폴더를 하나 만들어서 관리하면 어떨까요? (추가로 fixture로 관리할 수 있는 다른 데이터는 없는지 한번 살펴봅시다!)

Ubermensch0608 replied:
Fixture 폴더로 나누어 조금 더 깔끔하게 관리할 수 있게 해봤습니다.

daadaadaah commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,34 @@
+
+  const updateCheckedCategory = (checkedCategoryId) => {
+    const updatedCategories = categories.map((category) => {
+      const newCategory = {
+        ...category,
+        checked: false,
+      };
```
이렇게 하시는 이유가 API에는 `checked`가 없어서 그러신가요?

Ubermensch0608 replied:
네 데이터를 프론트에서 사용하기 쉽게 하려고 가공한 것인데 혹시 이런 방식이 데이터 관리에 좋지 않을까요?

daadaadaah replied:
아니요! 그냥 궁금해서 물어봤습니다. redux에서 관리하신 걸로 리팩토링 하신 것도 잘하셨어요!👍

daadaadaah commented on src/App.jsx:
```diff
@@ -1,5 +1,85 @@
+    setCategories(updatedCategories);
+  };
+
+  useEffect(() => {
+    (async () => {
+      Promise.all([
+        fetch('https://eatgo-customer-api.ahastudio.com/regions'),
+        fetch('https://eatgo-customer-api.ahastudio.com/categories'),
+      ]).then(async (response) => {
+        const initialRegions = await response[0].json();
+        const initialCategories = await response[1].json();
+
+        setRegions(initialRegions);
+        setCategories(initialCategories);
+      });
+    })();
+  }, []);
+
+  useEffect(() => {
+    const checkedRegion = regions.find((region) => region.checked);
+    const checkedCategory = categories.find((category) => category.checked);
+
+    if (checkedCategory && checkedRegion) {
+      (async () => {
+        const regionName = checkedRegion.name;
+        const categoryId = checkedCategory.id;
+
+        const response = await fetch(`https://eatgo-customer-api.ahastudio.com/restaurants?region=${regionName}&category=${categoryId}`);
+        const restaurantList = await response.json();
+
+        setRestaurants(restaurantList);
+      })();
+    }
+  }, [regions, categories]);
```
이부분은 redux와 redux-thunk를 사용해서 좀더 간단한 코드로 바꿔봅시다!

Ubermensch0608 replied:
네 아직 redux와 thunk이용은 자유롭게 하기 어렵지만 노력해보고있습니다!

daadaadaah reviewed (changes requested):
아직 완성하지 못했지만 먼저 PR 날리신 점 너무 잘하셨어요!👏
- TDD를 지금 연습하고 있는 단계이니 의도적으로 commit 메시지도 `XX 테스트 - XX 구현 코드 - XX (리팩토링)` 등의 형식으로 작성해보시는거 어떠세요? 또한, commit 작성하실 때, 작업단위를 좀더 작게 쪼개보는 연습을 하는게 좋을 것 같아요!
- 지금 작성해주신 https://github.com/CodeSoom/react-week5-assignment-1/pull/125/commits/c61de612b2658cda87661324121ccc8ffee625d6 을 보면, `RegionsContainer 테스트 코드 작성`이라고 되어 있는데, 테스트 코드 뿐만 아니라 `RegionsContainer` 구현 코드도 들어가 있고, 다른 코드들도 들어가 있어요!

Ubermensch0608 committed:
Refactor: 테스트 코드에서 fixture 분리

테스트 코드를 조금 더 쉽게 읽히게 하기위해 설명문을 다시 작성해보았습니다.
반복되는 fixture들은 따로 분리하여 재사용 및 관리가 좋아지도록 하였습니다.

Ubermensch0608 committed:
Refactor: 관심사에 해당하는 container로 state 관리

App에서 모든 state관리를 담당하는 것은 관심사의 분리에 맞지 않기 때문에 각 관심사에 해당하는 container로 비즈니스 로직을 옮겼습니다.

Ubermensch0608 committed:
지역 리스트 변경 액션 테스트

fixture를 이용해 지역 리스트 테스트 변경 테스트

Ubermensch0608 committed:
regions state관리 리팩토링

지역 데이터 state를 지역적으로 관리하던 것을 redux를 사용하여 전역으로 변경하였습니다.

Ubermensch0608 committed:
지역 리스트 비동기 fetching 리팩터링

redux-thunk를 이용하여 redux내에서 비동기 처리를 할 수 있도록 개선하였습니다.

daadaadaah commented on src/fixtures/restaurants.js:
```diff
@@ -0,0 +1,3 @@
+const restaurants = [{ id: 1, name: '야미야미' }];
```
여러개라는 의도가 드러나게 2~3개 정도 데이터로 만드시는 걸 추천드립니다.

Ubermensch0608 replied:
넵 다른 과제에서도 그렇게 적용해보도록 하겠습니다~

daadaadaah commented on src/App.jsx:
```diff
@@ -1,5 +1,34 @@
 export default function App() {
+  const [restaurants, setRestaurants] = useState([]);
+
+  // useEffect(() => {
+  //   const checkedRegion = regions.find((region) => region.checked);
+  //   const checkedCategory = categories.find((category) => category.checked);
+
+  //   if (checkedCategory && checkedRegion) {
+  //     (async () => {
+  //       const regionName = checkedRegion.name;
+  //       const categoryId = checkedCategory.id;
+
+  //       const response = await fetch(`https://eatgo-customer-api.ahastudio.com/restaurants?region=${regionName}&category=${categoryId}`);
+  //       const restaurantList = await response.json();
+
+  //       setRestaurants(restaurantList);
+  //     })();
+  //   }
+  // }, []);
```
나중에 주석 내용이 필요하더라도, PR 날리실 때는 정리해서 날리시는 걸 추천드려요!

Ubermensch0608 replied:
넵 주석 같이 필요 없는 데이터는 최대한 정리 후 보내도록 하겠습니다!

daadaadaah reviewed (changes requested):
계속 진행해주세요!
https://github.com/CodeSoom/react-week5-assignment-1/pull/125/files#r1060595357 이거 답변 부탁드려요!ㅎㅎ

Ubermensch0608 committed:
Refactor: 지역 state 관리를 전역으로 변경

비즈니스 로직을 Cotainer 컴포넌트에서 관장하고, 리덕스를 이용하여 전역으로 관리할 수 있도록 하였습니다.

Ubermensch0608 committed:
Refactor: 오류 테스트 코드 리팩토링

기능을 변경하면서 나타났던 오류 수정하였습니다.

Ubermensch0608 committed:
Refactor: 카테고리 컴포넌트 리팩토링

카테고리 컴포넌트 렌더링과 액션 함수 실행 여부를 확인하는 테스트 코드를 추가하였습니다.

Ubermensch0608 committed:
Refactor: 지역 목록 체크 액션 테스트 리팩터링

daadaadaah commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,40 @@
+    });
+  });
+
+  context('지역과 카테고리를 모두 체크했을 때', () => {
+    it('레스토랑 목록을 화면에 렌더한다.', () => {
+      useSelector.mockImplementation((selector) => selector({
+        restaurants,
+        categories: [],
+        regions: [],
+      }));
+
+      render(<RestaurantsContainer />);
+    });
+  });
```
expect가 빠져있네요?ㅎㅎ

daadaadaah commented on src/store/reducer.js:
```diff
@@ -13,6 +14,70 @@ const reducer = (state = initialState, action) => {
+    });
+
+    return {
+      ...state,
+      categories: newCategories,
+    };
+  }
+
+  if (action.type === 'checkRegion') {
+    const { regions } = state;
+    const { checkedRegionId } = action.payload;
+
+    const newRegions = regions.map((region) => {
+      const newRegion = {
+        ...region,
+        checked: false,
+      };
+
+      if (region.id === checkedRegionId) {
+        newRegion.checked = true;
+      }
+
+      return newRegion;
+    });
+
+    return {
+      ...state,
+      regions: newRegions,
+    };
+  }
+
+  if (action.type === 'setRestaurants') {
+    const { restaurants } = action.payload;
+
+    return {
+      ...state,
+      restaurants,
+    };
+  }
+
```
앞에 과제들에서 if문이 많아졌을 때, 어떤걸 활용해서 코드를 간단히 했었나요?

daadaadaah reviewed (changes requested):
조금씩 PR 날려주시는 모습 너무 좋습니다!👏
다음 PR 기다리고 있을게요~
